prow/cmd/deck/trace.go

```diff
@@ -0,0 +1,199 @@
+
+type lineData struct {
+	TimeStr   string `json:"time"`
+	time      time.Time
+	PrNum     int    `json:"pr"`
+	Repo      string `json:"repo"`
+	Org       string `json:"org"`
+	EventGUID string `json:"event-GUID"`
+}
+
+type podLine struct {
+	actual      []byte
+	unmarshaled lineData
+}
+
+// linesByTimestamp sorts pod lines by timestamp.
+// Useful when collecting logs across multiple pods.
+type linesByTimestamp []podLine
+
+// linesByTimestamp implements the sort.Interface interface.
+var _ sort.Interface = linesByTimestamp{}
+
+func (l linesByTimestamp) Len() int      { return len(l) }
+func (l linesByTimestamp) Swap(i, j int) { l[i], l[j] = l[j], l[i] }
+func (l linesByTimestamp) Less(i, j int) bool {
+	return l[i].unmarshaled.time.Before(l[j].unmarshaled.time)
+}
+
+// linesByTimestamp implements the fmt.Stringer interface.
+var _ fmt.Stringer = linesByTimestamp{}
+
+// Return valid json.
+func (pl linesByTimestamp) String() string {
+	sort.Sort(pl)
+
+	var log string
+	for i, line := range pl {
+		switch i {
+		case len(pl) - 1:
+			log += string(line.actual)
+		default:
+			// buf.ReadBytes('\n') does not remove the newline
+			log += fmt.Sprintf("%s,\n", strings.TrimSuffix(string(line.actual), "\n"))
+		}
+	}
+
+	return fmt.Sprintf("[%s]", log)
+}
+
+func handleTrace(ja *JobAgent) http.HandlerFunc {
+	return func(w http.ResponseWriter, r *http.Request) {
+		w.Header().Set("Cache-Control", "no-cache")
+		w.Header().Set("Access-Control-Allow-Origin", "*")
+
+		prNum, err := validateTraceRequest(r)
+		if err != nil {
+			logrus.Debugf("Invalid request: %v", err)
+			http.Error(w, fmt.Sprintf("Invalid request: %v", err), http.StatusBadRequest)
+			return
+		}
+
+		var targets []kube.Pod
+		for _, selector := range ja.c.Config().Deck.TraceTargets {
+			pods, err := ja.pkc.ListPods(selector)
+			if err != nil {
+				logrus.Debugf("Cannot list pods with selector %q: %v", selector, err)
+				http.Error(w, fmt.Sprintf("Cannot list pods with selector %q: %v", selector, err), http.StatusBadGateway)
+				return
+			}
+			for _, pod := range pods {
+				if pod.Status.Phase != kube.PodRunning {
+					logrus.Debugf("Ignoring pod %q: not in %s phase (phase: %s, reason: %s)",
+						pod.Metadata.Name, kube.PodRunning, pod.Status.Phase, pod.Status.Reason)
+					continue
+				}
+				targets = append(targets, pod)
+			}
+		}
+
+		if len(targets) == 0 {
+			logrus.Debug("No targets found.")
+			fmt.Fprint(w, "[]")
+			return
+		}
+
+		pr := r.URL.Query().Get(github.PrLogField)
+		repo := r.URL.Query().Get(github.RepoLogField)
+		org := r.URL.Query().Get(github.OrgLogField)
+		eventGUID := r.URL.Query().Get(github.EventGUID)
+
+		log := make(linesByTimestamp, 0)
+		var buf *bytes.Buffer
+		for _, pod := range targets {
+			// TODO: Cache this and use "since" as a pod/log url parameter to fetch
+			// newer logs.
+			podLog, err := ja.pkc.GetLog(pod.Metadata.Name)
+			if err != nil {
+				logrus.Debugf("cannot get logs from %q: %v", pod.Metadata.Name, err)
```

Comment:
Would be nice to somehow expose on the front-end that the view someone is looking at is incomplete if we run into a gathering or parsing issue

Reply:
Hrm, yeah. We can potentially add it in the log we return inside the endpoint. Let's tackle this in a follow-up.

Reply:
ACK

Reply:
https://github.com/kubernetes/test-infra/issues/5823